src/workerd/api/sockets.c++

```diff
@@ -271,8 +272,53 @@ jsg::Promise<void> Socket::close(jsg::Lock& js) {
     return abortPromise.then(js, [this](jsg::Lock& js) {
       resolveFulfiller(js, kj::none);
       return js.resolvedPromise();
-    }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
-  }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
+    }, [this](jsg::Lock& js, jsg::Value err) {
+      errorHandler(js, kj::mv(err));
+      return js.resolvedPromise();
+    });
+  }, [this](jsg::Lock& js, jsg::Value err) {
+    errorHandler(js, kj::mv(err));
+    return js.resolvedPromise();
+  });
+}
+
+jsg::Promise<void> Socket::closeImplNew(jsg::Lock& js) {
+  if (isClosing) {
+    return closedPromiseCopy.whenResolved(js);
+  }
+
+  isClosing = true;
+
+  // Wait until the socket connects (successfully or otherwise)
+  return openedPromiseCopy.whenResolved(js).then(js, [this](jsg::Lock& js) {
+    auto flushPromise = js.resolvedPromise();
+    if (!writable->getController().isClosedOrClosing()) {
+      flushPromise = writable->getController().flush(js);
+    }
+    return flushPromise;
```

Comment:
It's a bit wasteful to create the resolved promise only to immediately throw it away.

```suggestion
    if (!writable->getController().isClosedOrClosing()) {
      return writable->getController().flush(js);
    } else {
      return js.resolvedPromise();
    }
```